pkg/csilvm/server.go

```diff
@@ -1196,10 +1195,23 @@ func volumeOptsFromParameters(in map[string]string) (opts []lvm.CreateLogicalVol
 //
 // See https://jira.mesosphere.com/browse/DCOS_OSS-4642
 func SerializingInterceptor() grpc.UnaryServerInterceptor {
-	var lk sync.Mutex
+	// Instead of a mutex, use a weighted semaphore because it's sensitive to context cancellation and/or deadline
+	// expiration, which is important for maintaining a healthy request queue, and also helps prevent execution of
+	// operations that the calling CO is no longer interested in.
+	sem := semaphore.NewWeighted(1)
 	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
-		lk.Lock()
-		defer lk.Unlock()
+		err := sem.Acquire(ctx, 1)
+		if err != nil {
+			return nil, err
+		}
+		// Acquire can still succeed if the context is canceled, double-check it.
```

Comment:
Isn't it easier to just acquire the mutex, check `ctx.Done()` and return if that returns? Why do we have a semaphore?

Reply:
`Acquire` stops blocking if/when the context is cancelled. A `Mutex` doesn't do that. Consider that the request queue is full because the `request-limit` of 10 is reached. One of the calling contexts for a request "R8" in the queue expires. With a mutex, the queue remains full until the request "R1" being processed has completed. With a semaphore, request "R8" is dropped from the queue and makes room for another request "R11" to queue up.

This PR is about the interaction between the `SerializingInterceptor` (which forces (unordered) queues to form) and `RequestLimitInterceptor` (which limits queue length). If we used a mutex, "R11" would not be able to join the queue even though "R8" was cancelled by the caller (as long as "R1" was still in-flight).

Reply:
I can add a test case that shows this.

Reply:
That makes sense, thanks for he explanation. LGTM

Reply:
That makes sense, thanks for he explanation. LGTM